File: Programas en Java/Tienda online/ClienteTienda.java
import java.util.Scanner;

public class ClienteTienda {
    private Scanner sc; // Escaner para cada cliente y solo se accede desde aquí
    private CarritoCompra carrito; // Cada cliente tiene su propio carrito de compra

    public ClienteTienda() { // Clase para cada cliente
        sc = new Scanner(System.in); // Se crea un objeto de tipo Scanner
        carrito = new CarritoCompra();
    }

    public void iniciarMenu() { // Método para inciar el menú del cliente
        int op = 0;

        do  {
            mostrarMenu();
            op = leerOpcion();

            switch (op) {
                case 1:
                    buscarArticulo();
                    break;
                case 2:
                    listarArticulos();
                    break;
                case 3:
                    agregarAlCarrito();
                    break;
                case 4:
                    editarCarrito();
                    break;
                case 5:
                    finalizarCompra();
                    break;
                default:
                    System.out.println(">> ERROR: La opción no es valida");
                    break;
            }
        } while (op != 5);
        
        sc.close();
    }

    private void mostrarMenu() { // Método para mostrar el menú del cliente
        System.out.println("/*------------------.");
        System.out.println("| Tienda: LalitoXDE |");
        System.out.println("`------------------*/");
        System.out.println("\n>> Elije una de las opciones");
        System.out.println("\n1.- Buscar artículos");
        System.out.println("2.- Listar artírculos por tipo");
        System.out.println("3.- Agregar artículos al carrito de compra");
        System.out.println("4.- Editar el contenido del carrito");
        System.out.println("5.- Finalizar compra");
    }

    private int leerOpcion() { // Método para leer la opción
        System.out.print("\nOpción: ");
        return sc.nextInt();
    }

    private void buscarArticulo() { // Método para buscar un artículo
        System.out.println("/*-----------------.");
        System.out.println("| BUSCAR ARTÍCULOS |");
        System.out.println("`-----------------*/");
    }

    private void listarArticulos() { // Método para listar los artículos
        System.out.println("/*-----------------.");
        System.out.println("| LISTAR ARTÍCULOS |");
        System.out.println("`-----------------*/");
    }

    private void agregarAlCarrito() { // Método para agregar un artículo al carrito
        System.out.println("/*-----------------------------.");
        System.out.println("| AGREGAR ARTÍCULOS AL CARRITO |");
        System.out.println("`-----------------------------*/");
    }

    private void editarCarrito() { // Método para editar el carrito de compra
        System.out.println("/*--------------------------------.");
        System.out.println("| EDITAR EL CONTENIDO DEL CARRITO |");
        System.out.println("`--------------------------------*/");
    }

    private void finalizarCompra() { // Método para finalizar la compra
        System.out.println("\n>> La compra ha finalizado con éxito, su comprobante se muestra a continuación:");
    }

    public static void main(String[] args) { // Clase principal
        ClienteTienda cliente = new ClienteTienda(); // Se crea un nevo cliente
        cliente.iniciarMenu(); // Se muestra el menú para el cliente
    }   
}

File: Programas en Python/Tienda en línea/V4/ClienteTienda.py
import json, socket
from FuncionesCliente import listarArticulos, limpiarTerminal, mostrarBusqueda
from FuncionesCliente import mostrarMensaje, mostrarCarrito, esperarTecla

# Cremos al cliente y lo conectamos al servidor
cliente = socket.socket(socket.AF_INET, socket.SOCK_STREAM) # Creamos al cliente (dirección IPv4, protocolo TCP)
cliente.connect(("127.0.0.1", 5000)) # Conectamos al cliente con el servidor (dirección IP, puerto)

# Menú del cliente
while (True): # Mientras no se cierre el cliente
    limpiarTerminal()
    print("""
/*---------------------------.
| TIENDA EN LÍNEA: LALITOXDE |
`---------------------------*/

>> Elije una de las opciones
      
1.- Listar artículos de la tienda
2.- Buscar un artículo
3.- Carrito de compras
4.- Finalizar la compra
5.- Salir de la tienda
""")
    
    opcion = input("Opción: ")

    match (opcion):
        case '1':
            print("""
/*--------------------------------.
| LISTA DE ARTÍCULOS EN LA TIENDA |                  
`--------------------------------*/
""")
            solicitud = {"accion": "LISTAR_ARTICULOS"} # Creamos la solicitud como un diccionario
            cliente.send(json.dumps(solicitud).encode("utf-8")) # Enviamos al servidor la solicitud serializada
            datosRecibidos = cliente.recv(4096).decode() # Recibe y deserializa los datos recibidos desde el servidor

            articulos = json.loads(datosRecibidos) # Convierte los datos recibidos en JSON a un diccionario

            listarArticulos(articulos) # Lista los artículos de la tienda

            # Esperamos una tecla
            print("\n>> Presiona una tecla para continuar...")
            tecla = esperarTecla()

        case '2':
            print("""
/*-------------------.
| BUSCAR UN ARTÍCULO |                  
`-------------------*/
""")
            buscar = input("Escribe el nombre o la marca del/los artículo(s) a buscar: ")

            solicitud = {"accion": "BUSCAR_ARTICULOS", "buscar": buscar} # Creamos la solicitud como un diccionario
            cliente.send(json.dumps(solicitud).encode("utf-8")) # Enviamos al servidor la solicitud serializada
            datosRecibidos = cliente.recv(4096).decode() # Recibe y deserializa los datos recibidos desde el servidor

            resultadoBusqueda = json.loads(datosRecibidos) # Convierte los datos recibidos en JSON a un diccionario

            if ("mensaje" in resultadoBusqueda): # No hubo coincidencias
                mostrarMensaje(resultadoBusqueda)
            else: # Si hubo al menos una coincidencia
                mostrarBusqueda(resultadoBusqueda) # # Se muestran los artículos encontrados
            
            # Esperamos una tecla
            print("\n>> Presiona una tecla para continuar...")
            tecla = esperarTecla()

        case '3':
            # Menú para el carrito
            while (True):
                limpiarTerminal()
                print("""
/*-------------------.
| CARRITO DE COMPRAS |
`-------------------*/
                  
>> Elije una de las opciones

1.- Ver carrito
2.- Agregar un artículo al carrito
3.- Eliminar un artículo del carrito
4.- Volver
""")
                
                opcion = input("Opción: ")

                match (opcion):
                    case '1':
                        print("""
/*------------------------.
| ARTÍCULOS EN EL CARRITO |                              
`------------------------*/
""")
                        solicitud = {"accion": "MOSTRAR_CARRITO"} # Creamos la solicitud como un diccionario
                        cliente.send(json.dumps(solicitud).encode("utf-8")) # Enviamos al servidor la solicitud serializada
                        datosRecibidos = cliente.recv(4096).decode() # Recibe y deserializa los datos recibidos desde el servidor

                        articulosCarrito = json.loads(datosRecibidos) # Convierte los datos recibidos en JSON a un diccionario

                        if ("mensaje" in articulosCarrito): # El carrito esta vacío
                            mostrarMensaje(articulosCarrito)
                        else: # Si hay al menos un artículo en el carrito
                            mostrarCarrito(articulosCarrito)

                        # Esperamos una tecla
                        print("\n>> Presiona una tecla para continuar...")
                        tecla = esperarTecla()

                    case '2':
                        print("""
/*-----------------.
| AGREGAR ARTÍCULO |
`-----------------*/
""")
                        agregar = input("Escribe el id o el nombre del artículo a agregar: ")
                        cantidad = input("Escribe la cantidad a agregar de ese artículo: ")

                        solicitud = {"accion": "AGREGAR_CARRITO", "articulo": buscar, "cantidad": int(cantidad)} # Creamos la solicitud como un diccionario
                        cliente.send(json.dumps(solicitud).encode("utf-8")) # Enviamos al servidor la solicitud serializada

                        # Esperamos una tecla
                        print("\n>> Presiona una tecla para continuar...")
                        tecla = esperarTecla()

                    case '3':
                        print("""
/*------------------.
| ELIMINAR ARTÍCULO |            
`------------------*/
""")
                        eliminar = input("Escribe el id o el nombre del artículo a eliminar: ")
                        cantidad = input("Escribe la cantidad a eliminar de ese artículo: ")

                        solicitud = {"accion": "ELIMINAR_CARRITO", "articulo": eliminar, "cantidad": int(cantidad)} # Creamos la solicitud como un diccionario
                        cliente.send(json.dumps(solicitud).encode("utf-8")) # Enviamos al servidor la solicitud serializada

                        # Esperamos una tecla
                        print("\n>> Presiona una tecla para continuar...")
                        tecla = esperarTecla()

                    case '4':
                        break

                    case _:
                        print("\n>> La opción no es válida")

                        # Esperamos una tecla
                        print("\n>> Presiona una tecla para continuar...")
                        tecla = esperarTecla()

        case '4':
            print("""
/*--------------------.
| FINALIZAR LA COMPRA |                  
`--------------------*/
""")
            solicitud = {"accion": "FINALIZAR_COMPRA"} # Creamos la solicitud como un diccionario

            # Esperamos una tecla
            print("\n>> Presiona una tecla para continuar...")
            tecla = esperarTecla()

        case '5':
            print("\n>> Gracias por visitar nuestra tienda, esperamos volverte a ver")
            cliente.close()
            break

        case _:
            print("\n>> La opción no es válida")

            # Esperamos una tecla
            print("\n>> Presiona una tecla para continuar...")
            tecla = esperarTecla()